### PR TITLE
refactor regression ops to nnvm interface

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
refactor regression operators to nnvm interface (then I will add sparse support for these operators in a separate PR later).

Fix the docs. Please see https://github.com/apache/incubator-mxnet/issues/9528.

Add unittest for `MAERegressionOutput`.

cc @eric-haibin-lin @sxjscience 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] refactor, existing unit tests for `LinearRegressionOutput` and `LogisticRegressionOutput`. Add unittest for `MAERegressionOutput`.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/regression_output.cc [hunk 1]
```diff
@@ -106,13 +135,40 @@ It is suitable for binary classification or probability prediction tasks.
...
+.set_num_inputs(2)
+.set_num_outputs(2)
+.set_attr_parser(ParamParser<RegressionOutputParam>)
+.set_attr<nnvm::TIsBackward>("TIsBackward", true)
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+[](const NodeAttrs& attrs){
+  // inputs are in_label and out_data, outputs are data_grad and label_grad
+  return std::vector<std::pair<int, int> >{{1, 0}};
+})
+.set_attr<FCompute>("FCompute<cpu>", RegressionBackward<cpu, mshadow_op::minus>);
```
Comment (sxjscience) [inline src/operator/regression_output.cc]: We can consider creating a macro (like in https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/elemwise_binary_op.h#L554-L569) for defining regression OPs as most of the codes are similar.

File: src/operator/regression_output-inl.h [hunk 2]
```diff
@@ -50,146 +50,90 @@ struct RegressionOutputParam : public dmlc::Parameter<RegressionOutputParam> {
...
+template<typename xpu, typename ForwardOp>
+void RegressionForward(const nnvm::NodeAttrs& attrs,
+                       const OpContext& ctx,
+                       const std::vector<TBlob>& inputs,
+                       const std::vector<OpReqType>& req,
+                       const std::vector<TBlob>& outputs) {
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  MSHADOW_REAL_TYPE_SWITCH(inputs[reg_enum::kData].type_flag_, DType, {
+    MXNET_ASSIGN_REQ_SWITCH(req[reg_enum::kOut], Req, {
+      DType* in_data = inputs[reg_enum::kData].dptr<DType>();
```
Comment (eric-haibin-lin) [inline src/operator/regression_output-inl.h]: nit: const

File: src/operator/regression_output-inl.h [hunk 3]
```diff
@@ -50,146 +50,90 @@ struct RegressionOutputParam : public dmlc::Parameter<RegressionOutputParam> {
...
+                        const std::vector<TBlob>& inputs,
+                        const std::vector<OpReqType>& req,
+                        const std::vector<TBlob>& outputs) {
+  const RegressionOutputParam& param = nnvm::get<RegressionOutputParam>(attrs.parsed);
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  // inputs are in_label, out_data
+  // outputs are data_grad, label_grad
+  MSHADOW_REAL_TYPE_SWITCH(inputs[1].type_flag_, DType, {
+    MXNET_ASSIGN_REQ_SWITCH(req[0], Req, {
+      DType* in_label = inputs[0].dptr<DType>();
```
Comment (eric-haibin-lin) [inline src/operator/regression_output-inl.h]: nit: const

File: src/operator/regression_output.cc [hunk 4]
```diff
@@ -18,86 +18,115 @@
...
 
-:math:`\text{SquaredLoss}(y, \hat{y} ) = \frac{1}{n} \sum_{i=0}^{n-1} \left( y_i - \hat{y}_i \right)^2`
+:math:`\text{SquaredLoss}(\textbf{Y}, \hat{\textbf{Y}} ) = \frac{1}{n} \sum_{i=0}^{n-1} \lVert  \textbf{y}_i - \hat{\textbf{y}}_i  \rVert_2`
 
 .. note::
    Use the LinearRegressionOutput as the final output layer of a net.
 
-By default, gradients of this loss function are scaled by factor `1/n`, where n is the number of training examples.
-The parameter `grad_scale` can be used to change this scale to `grad_scale/n`.
+By default, gradients of this loss function are scaled by factor `1/m`, where m is the number of dimensions of a training example.
```
Comment (eric-haibin-lin) [inline src/operator/regression_output.cc]: `number of dimensions ` is a bit confusing to me. For shape `(3,4)`, my first thought of num dimension is two (2-D shape). Is "number of features" clearer?

File: src/operator/regression_output.cc (hunk 1 above)
Comment (eric-haibin-lin) [reply src/operator/regression_output.cc]: +1

**Comment by ZiyueHuang** [comment]
Thanks for the comments. Updated.

File: src/operator/regression_output.cc [hunk 5]
```diff
@@ -18,86 +18,96 @@
...
 
-:math:`\text{MAE}(y, \hat{y} ) = \frac{1}{n} \sum_{i=0}^{n-1} \left| y_i - \hat{y}_i \right|`
+:math:`\text{MAE}(\textbf{Y}, \hat{\textbf{Y}} ) = \frac{1}{n} \sum_{i=0}^{n-1} \lVert \textbf{y}_i - \hat{\textbf{y}}_i \rVert_1`
 
 .. note::
    Use the MAERegressionOutput as the final output layer of a net.
 
-By default, gradients of this loss function are scaled by factor `1/n`, where n is the number of training examples.
-The parameter `grad_scale` can be used to change this scale to `grad_scale/n`.
+By default, gradients of this loss function are scaled by factor `1/m`, where m is the number of features of a training example.
```
Comment (sxjscience) [inline src/operator/regression_output.cc]: I think the word `features` is not very appropriate. We can use `regression outputs`

**Comment by sxjscience** [comment]
Added some minor docstring comments. The PR looks good.

File: src/operator/regression_output.cc (hunk 5 above)
Comment (ZiyueHuang) [reply src/operator/regression_output.cc]: Updated

**Comment by ZiyueHuang** [comment]
Seems that I mistakenly checkout that test file... Added back.

**Pull request merged**